runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MetricUpdatesProtos.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.runners.core.metrics;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.BiConsumer;
+import org.apache.beam.model.pipeline.v1.MetricsApi.MonitoringInfo;
+import org.apache.beam.sdk.metrics.MetricKey;
+
+/** Convert {@link MetricUpdates} to a {@link List} of {@link MonitoringInfo}. */
+public class MetricUpdatesProtos {
+  public static List<MonitoringInfo> toProto(MetricUpdates metricUpdates) {
+    ArrayList<MonitoringInfo> monitoringInfos = new ArrayList<>();
+
+    fromMetrics(
```

Comment:
Similarily in this code, the BiConsumer usage complicates this and makes it harder to read.
Its simple enough to
(1) make a for loop for each of these calls, iterating through the MetricUpdates
(2) call a helper to do everything else

A lot of this problem comes from the fact that MetricUpdate is just a generic class with no knowledge of what its actually storing. So you can't take advantage of any polymophism with a MetricGaugeUpdate, MetricDistributionUpdate, MetricCounterUpdate, etc. So a lot of the code surrounding this special cases or writes switches for each one. If this polymorphism existed you could tell the MetricUpdate to set itself on a MonitoringInfo/SimpleMonitoringInfoBuilder. That's an option as well, but it will require a larger change. Better to table for now.

A lot of the cases where you pass in a callback seem to be for similar reasons.